OSU-CS/automod
# Port to new slack library and add API

solves #36 

### Description

Switch from use of deprecated RTM client to the new bolt-python library and make it easier for new folks to contribute. This includes:
 - switch to running as a web API that receives events at `/slack/events`
 - supporting deployment configurations
 - manifest.yml for setting up new apps for testing (we can try to convert the prod deployment app over but may end up needing to create a new one) 
 - documentation of how to set up a development environment now that legacy tokens are no longer available using the new recommended workflow from slack

### Acceptance Criteria

 - [ ] a fresh workspace integration can be set up following the docs
 - [ ] FastAPI docs should load on `/docs` with a single endpoint for POST `/slack/events`
 - [ ] emoji events are logged to:
   - [ ] `#admin` immediately
   - [ ] `#emoji_meta` on schedule
 - [ ] large batches of emoji:
   - [ ] do not get hit too hard with the rate limiter (some failures are expected per current behavior)
   - [ ] do not crash the app when an error is hit
 - [ ] new channel is logged to `#admin` immediately
 - [ ] new user is logged to `#admin_logs` immediately


### Testing

1. Clone a fresh repo to follow/test the development setup instructions
2. Follow the integration setup instructions to test setting up a workspace/application (can skip actual _creation_ of a workspace if you already have a personal test workspace)
3. Test the deployment instructions for ngrok. 
4. When complete, docs endpoint should load for ngrok url. 
5. In the slack app, test ngrok url deployment by adding and verifying the request_url successfully.
6. Repeat steps 3-5 for heroku deployment.
7. Go into your test slack workspace.
8. Add and remove emoji and verify that it hits the `#admin` channel.
9. Add a very large number of emoji using the [Bulk Emoji Uploader chrome extension](https://chrome.google.com/webstore/detail/neutral-face-emoji-tools/anchoacphlfbdomdlomnbbfhcmcdmjej?hl=en).
10. Add a new channel and verify that it hits the `#admin` channel.
11. Invite yourself as a new user to the workspace (pro tip - add +sometext to your gmail for unique addresses to invite) and verify that it hits the `#admin_logs` channel. 
12. Check back later to verify that emoji reports have hit `#emoji_meta`.

### Additional Notes

This change required full testing of all existing features of the app. The only potential regression I have found so far is the behavior where automod's icon matches the emoji being added. I have added `chat:write.customize` permissions to support this but have not yet tested. If it doesn't Just Work :tm: from this permission, I think it should be filed as a new issue as it will require deeper investigation.
